File: java/src/ThirdMaximumNumber.java
public class ThirdMaximumNumber {
    public static int thirdMax(int[] array) {
        if (array.length == 1) return array[0];
        if (array.length == 2) return Math.max(array[0], array[1]);
        long max = array[0], secondMax = Long.MIN_VALUE, thirdMax = Long.MIN_VALUE;
        for (int index = 1 ; index < array.length ; index++) {
            if (array[index] > max) {
                thirdMax = secondMax;
                secondMax = max;
                max = array[index];
            } else if (array[index] > secondMax && array[index] < max) {
                thirdMax = secondMax;
                secondMax = array[index];
            } else if (array[index] > thirdMax&& array[index] < secondMax) {
                thirdMax = array[index];
            }
        }
        return (int) (thirdMax == Long.MIN_VALUE ? max : thirdMax);
    }
}

File: python/third_maximum_number.py
from typing import List


class Solution:
    def thirdMax(self, array: List[int]) -> int:
        if len(array) == 1: return array[0]
        if len(array) == 2: return max(array[0], array[1])
        infinity = float('inf')
        first, second, third = array[0], -infinity, -infinity
        index = 1
        while index < len(array):
            if array[index] > first:
                third = second
                second = first
                first = array[index]
            elif second < array[index] < first:
                third = second
                second = array[index]
            elif third < array[index] < second:
                third = array[index]
            index += 1
        return first if third == -infinity else third
